pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +372,422 @@ pub mod pallet {
+		) -> bool {
+			if T::Investment::accepted_payment_currency(investment_id, currency) {
+				true
+			} else {
+				T::PoolInspect::currency_for(investment_id.of_pool())
+					.map(|pool_currency| T::TokenSwaps::valid_pair(pool_currency, currency))
+					.unwrap_or(false)
+			}
+		}
+
+		fn accepted_payout_currency(
+			investment_id: T::InvestmentId,
+			currency: T::CurrencyId,
+		) -> bool {
+			if T::Investment::accepted_payout_currency(investment_id, currency) {
+				true
+			} else {
+				T::PoolInspect::currency_for(investment_id.of_pool())
+					.map(|pool_currency| T::TokenSwaps::valid_pair(currency, pool_currency))
+					.unwrap_or(false)
+			}
+		}
+	}
+
+	pub struct FulfilledSwapOrderHook<T>(PhantomData<T>);
+	impl<T: Config> StatusNotificationHook for FulfilledSwapOrderHook<T> {
+		type Error = DispatchError;
+		type Id = T::SwapId;
+		type Status = SwapOf<T>;
+
+		fn notify_status_change(
+			swap_id: T::SwapId,
+			last_swap: SwapOf<T>,
+		) -> Result<(), DispatchError> {
+			let (who, investment_id) =
+				SwapIdToForeignId::<T>::get(swap_id).ok_or(Error::<T>::SwapOrderNotFound)?;
+
+			let pending_amount = match T::TokenSwaps::get_order_details(swap_id) {
+				Some(swap) => swap.amount_in,
+				None => {
+					SwapIdToForeignId::<T>::remove(swap_id);
+					T::Balance::zero()
+				}
+			};
+
+			ForeignInvestmentInfo::<T>::mutate_exists(&who, investment_id, |maybe_info| {
+				if let Some(info) = maybe_info {
+					if info.increase_swap_id == Some(swap_id) {
+						T::Investment::update_investment(
+							&who,
+							investment_id,
+							T::Investment::investment(&who, investment_id)?
+								.ensure_add(last_swap.amount_in)?,
+						)?;
+
+						if pending_amount.is_zero() {
+							info.increase_swap_id = None;
+						}
+					} else if info.decrease_swap_id == Some(swap_id) {
+						T::DecreasedForeignInvestOrderHook::notify_status_change(
+							(who.clone(), investment_id),
+							ExecutedForeignDecreaseInvest {
+								amount_decreased: last_swap.amount_in,
+								foreign_currency: info.base.foreign_currency,
+								amount_remaining: pending_amount,
+							},
+						)?;
+
+						if pending_amount.is_zero() {
+							info.decrease_swap_id = None;
+						}
+					}
+				}
+
+				Ok::<_, DispatchError>(())
+			})?;
+
+			ForeignRedemptionInfo::<T>::mutate_exists(&who, investment_id, |maybe_info| {
+				if let Some(info) = maybe_info {
+					if info.swap_id == Some(swap_id) {
+						info.swapped_amount.ensure_add_assign(last_swap.amount_in)?;
+
+						if info.base.collected.amount_collected == info.swapped_amount {
+							let tranche_tokens = info.pending_tranche_tokens_to_send()?;
+							info.tranche_tokens_sent.ensure_add_assign(tranche_tokens)?;
+
+							T::CollectedForeignRedemptionHook::notify_status_change(
+								(who.clone(), investment_id),
+								ExecutedForeignCollect {
+									currency: info.base.foreign_currency,
+									amount_currency_payout: last_swap.amount_in,
```

Comment:
Why do we increment the `info.swapped_amount` but then only use `last_swap.amount_in` here? Shouldn't we use `info.swapped_amount`?

Reply:
`info.swapped_amount` is the total amount of all swaps done until this moment. It's only used to determine whether we've swapped the same amount as we've collected, in order to dispatch the `ExecutedForeignCollect` event.

But I think you're right regarding the fact that we should not use `last_swap.amount_in` because it is only the amount of the last swap. We should use the amount from the last `ExecutedForeignCollect`

Reply:
Fixed